src/bson.c

```diff
@@ -86,59 +88,86 @@ char* php_phongo_field_path_as_string(php_phongo_field_path* field_path)
 	return path;
 }
 
-php_phongo_field_path* php_phongo_field_path_alloc(void)
+php_phongo_field_path* php_phongo_field_path_alloc(bool free_elements)
 {
 	php_phongo_field_path* tmp = ecalloc(1, sizeof(php_phongo_field_path));
 	tmp->ref_count             = 1;
+	tmp->free_elements         = free_elements;
 
 	return tmp;
 }
 
 void php_phongo_field_path_free(php_phongo_field_path* field_path)
 {
+	if (field_path->free_elements) {
+		size_t i;
+
+		for (i = 0; i < field_path->size; i++) {
+			efree(field_path->elements[i]);
+		}
+	}
 	if (field_path->elements) {
 		efree(field_path->elements);
 	}
+	if (field_path->element_types) {
+		efree(field_path->element_types);
+	}
 	efree(field_path);
 }
 
 static void php_phongo_field_path_ensure_allocation(php_phongo_field_path* field_path, size_t level)
 {
-	if (level >= field_path->allocated_levels) {
+	if (level >= field_path->allocated) {
 		size_t i;
 
-		field_path->allocated_levels = field_path->current_level + PHONGO_FIELD_PATH_EXPANSION;
-		field_path->elements         = erealloc(field_path->elements, sizeof(char**) * field_path->allocated_levels);
+		field_path->allocated     = field_path->size + PHONGO_FIELD_PATH_EXPANSION;
+		field_path->elements      = erealloc(field_path->elements, sizeof(char**) * field_path->allocated);
+		field_path->element_types = erealloc(field_path->element_types, sizeof(php_phongo_bson_field_path_item_types*) * field_path->allocated);
 
-		for (i = level; i < field_path->allocated_levels; i++) {
-			field_path->elements[i] = NULL;
+		for (i = level; i < field_path->allocated; i++) {
+			field_path->elements[i]      = NULL;
+			field_path->element_types[i] = PHONGO_FIELD_PATH_ITEM_NONE;
 		}
 	}
 }
 
 void php_phongo_field_path_write_item_at_current_level(php_phongo_field_path* field_path, const char* element)
 {
-	php_phongo_field_path_ensure_allocation(field_path, field_path->current_level);
+	php_phongo_field_path_ensure_allocation(field_path, field_path->size);
+
+	if (field_path->free_elements) {
```

Comment:
If `free_elements` isn't just used for free logic, I might suggest calling it `copy_elements` instead. That makes it clearer that this function is doing to `estrdup()` and likewise makes sense when you check it to free each element later on.

As-is, I would expect the caller to do `estrdup()` on its own.

Reply:
I agree, and I had already updated this locally. I had changed it to "owns_elements" though — let me know which one you think is better.

Reply:
`owns_elements` works for me.